### PR TITLE
Fix issue #16

### PR DESCRIPTION
**Pull request by baranga**
Fix issue #16

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -98,6 +98,22 @@ class MongoTenant {
     return this.options.requireTenantId === true;
   }
 
+  /**
+   * Checks if instance is compatible to other plugin instance
+   *
+   * For population of referenced models it's necessary to detect if the tenant
+   * plugin installed installed in these models is compatible to the plugin of
```
Comment (alrik) [inline index.js]: _installed installed_

File: index.js [hunk 2]
```diff
@@ -190,23 +206,30 @@ class MongoTenant {
...
+          const boundDb = me.createBoundDb(Model.db, tenantId);
+          modelCache[tenantId] = me.createTenantAwareModel(Model, tenantId, boundDb);
+        }
 
-      return modelCache[tenantId];
-    };
+        return modelCache[tenantId];
+      },
+
+      getMongoTenantPluginInstance: function () {
```
Comment (alrik) [inline index.js]: what about making it a property?

```js
get mongoTenant() {
  return me;
}
```

File: index.js [hunk 3]
```diff
@@ -314,6 +342,35 @@ class MongoTenant {
...
+   * @param {*} tenantId
+   * @returns {Connection}
+   */
+  createBoundDb(unboundDb, tenantId) {
+    const me = this;
+    const boundDb = Object.create(unboundDb);
+    boundDb.model = (name) => {
+      const unboundModel = unboundDb.model(name);
+      const otherPluginPresent = unboundModel.schema.plugins.some(
+        plugin => (plugin.fn === mongoTenantPlugin)
```
Comment (alrik) [inline index.js]: i think the other plugin does not need to be the same module :thinking:
in scenarios where you split your models up into separate packages we might have different mongo-tenant references.

File: index.js [hunk 4]
```diff
@@ -314,6 +342,35 @@ class MongoTenant {
...
   }
 
+  /**
+   * Create db connection bound to a specific tenant
+   *
+   * @param {Connection} unboundDb
+   * @param {*} tenantId
+   * @returns {Connection}
+   */
+  createBoundDb(unboundDb, tenantId) {
```
Comment (alrik) [inline index.js]: I would go with `createTenantAwareDb` to have consistent naming.

File: index.js [hunk 5]
```diff
@@ -190,23 +206,30 @@ class MongoTenant {
...
+        // lookup tenant-bound model in cache
+        if (!modelCache[tenantId]) {
+          let Model = this.model(this.modelName);
 
-        // Cache the tenant bound model class.
-        modelCache[tenantId] = me.createTenantAwareModel(Model, tenantId);
-      }
+          // Cache the tenant bound model class.
+          const boundDb = me.createBoundDb(Model.db, tenantId);
+          modelCache[tenantId] = me.createTenantAwareModel(Model, tenantId, boundDb);
```
Comment (alrik) [inline index.js]: it feels like the db with tenant context should be created within `createTenantAwareModel`

File: index.js [hunk 6]
```diff
@@ -98,6 +98,22 @@ class MongoTenant {
...
+   *
+   * For population of referenced models it's necessary to detect if the tenant
+   * plugin installed installed in these models is compatible to the plugin of
+   * the host model. If they are compatible they are one the same "level".
+   *
+   * @param {MongoTenant} plugin
+   */
+  isCompatibleTo(plugin) {
+    return (
+      plugin instanceof MongoTenant &&
```
Comment (alrik) [inline index.js]: I think `instanceOf``is not safe to use in multi-package setups.

**Review by alrik — CHANGES_REQUESTED** [review]

**Comment by alrik** [comment]
Awesome @baranga :1st_place_medal: 

It might be part of another issue, but do you know what happens, when you save a cross tenant ref? 
Something like:

```js
const subSchema = new Schema({});
const parentSchema = new Schema({
  sub: [{ type: Schema.Types.ObjectId, ref: 'Sub' }]
});
subSchema.pluging(mongoTenant);
parentSchema.plugin(mongoTenant);
const Sub = mongoose.model('Sub', subSchema);
const Parent = mongoose.model('Person', parentSchema);

const sub = await Sub.byTenant('tenant-1').create({});
const parent = await Parent.byTenant('tenant-2').create({ sub: sub._id });
```

**Comment by baranga** [comment]
@alrik this will save the references in the database but don't populate them back if fetched in a tenant context. See https://github.com/craftup/node-mongo-tenant/blob/a04f92f7b0def5abf4cbfe616df9dd42e5e4a5af/test/middleware.js#L124

File: index.js (hunk 2 above)
Comment (baranga) [reply index.js]: Don't know if that works for schema statics.

Comment (baranga) [reply index.js]: It works :)

**Comment by baranga** [comment]
@alrik Incorporated all your changes. Could you review again please?

**Comment by sashahilton00** [comment]
Is there any reason that this never made it into master?

**Comment by alrik** [comment]
@sashahilton00 sorry for this very late response :crying_cat_face: 
There was not enough pressure behind this issue :)
This issue comes with some conceptual complexity and i did not took the time yet to test every potential side effect...
Do you need this feature implemented?

**Comment by sashahilton00** [comment]
Not a problem @alrik, we have used a couple of hacky workarounds, but it would be good to get this merged at somepoint as we can tidy up our codebase a little once it is.

**Comment by baranga** [comment]
@alrik I merged develop and fixed an infinite loop bug unveiled by the new tests cases from that merge.

**Comment by alrik** [comment]
Nice. I'll have a final look on it tomorrow.

**Comment by alrik** [comment]
@baranga I don't like to introduce a BC break :thinking: 
couldn't we define the schema property like this?
```js
constructor(schema, options) {
    this._modelCache = {};
    this.options = options || {};

    Object.defineProperty(this, 'schema', {
      get() {
        return schema;
      },
    });
}
``` 
This way schema prop will only be made un-enumerable (and thus hidden from the deep-merge).

**Comment by baranga** [comment]
@alrik the issue I tried to solve with the latest commit was a infinite loop happening on dump deep merge of two schemas. I wonder if two plugin instance should be merged at all :thinking:

So maybe it would be best to define all properties in the get style. But that may cause BC issues again. Sounds like we should either not expose anything in 2.0 or at least document that the plugin internals are subject of changes (which should be the default case any way).

**Review by alrik — APPROVED** [review]

**Pull request merged**